logbook-jaxrs/src/main/java/org/zalando/logbook/jaxrs/RemoteRequest.java

```diff
@@ -11,18 +12,95 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static java.util.Optional.ofNullable;
+import static org.zalando.fauxpas.FauxPas.throwingUnaryOperator;
+import static org.zalando.logbook.jaxrs.ByteStreams.toByteArray;
 
+@AllArgsConstructor
 final class RemoteRequest implements HttpRequest {
 
-    private final ContainerRequestContext context;
-    private byte[] body;
+    private interface State {
+
+        default State with() {
+            return this;
+        }
+
+        default State without() {
+            return this;
+        }
+
+        default State buffer(final ContainerRequestContext context) throws IOException {
+            return this;
+        }
+
+        default byte[] getBody() {
+            return new byte[0];
+        }
+
+    }
+
+    private static final class Unbuffered implements State {
+
+        @Override
+        public State with() {
+            return new Offering();
+        }
+
+    }
+
+    private static final class Offering implements State {
+
+        @Override
+        public State without() {
+            return new Unbuffered();
+        }
+
+        @Override
+        public State buffer(
+                final ContainerRequestContext context) throws IOException {
+
+            final byte[] body = toByteArray(context.getEntityStream());
+            context.setEntityStream(new ByteArrayInputStream(body));
+            return new Buffering(body);
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Buffering implements State {
+
+        private final byte[] body;
+
+        @Override
+        public State without() {
+            return new Ignoring(body);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return body;
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Ignoring implements State {
+
+        private final byte[] body;
+
+        @Override
+        public State with() {
+            return new Buffering(body);
+        }
 
-    public RemoteRequest(final ContainerRequestContext context) {
-        this.context = context;
     }
 
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)